src/v/iceberg/rest_client/catalog_client.h

```diff
@@ -0,0 +1,100 @@
+expected<json::Document> parse_json(iobuf&& raw_response);
+
+// The catalog client enables high level operations against a catalog rest api.
+// It also ensures that a token is available before making calls. A client is
+// lightweight and expected to be created for single operations or closely
+// related operations.
+class catalog_client {
+public:
+    /// \brief Construct a catalog client
+    /// \param client an object implementing the abstract client interface which
+    /// can perform an http request
+    /// \param endpoint the rest catalog server hostname to connect to,
+    /// including scheme
+    /// \param credentials credentials to be used to acquire oauth token from
+    /// catalog server
+    /// \param base_url an optional path to be prefixed to all API calls made by
+    /// this client
+    /// \param prefix an optional prefix to be added after the API version,
+    /// applied to all calls made by this client
+    /// \param api_version pi version used to construct URLs, defaults to v1
+    /// \param token an optional oauth token which will be used if valid. If
+    /// expired, a new one will be acquired
+    /// \param retry_policy a retry policy
+    /// used to determine how failing calls will be retried
+    catalog_client(
+      http::abstract_client& client,
+      std::string_view endpoint,
+      credentials credentials,
+      std::optional<std::string_view> base_url = std::nullopt,
+      std::optional<std::string_view> prefix = std::nullopt,
+      std::string_view api_version = "v1",
+      std::optional<oauth_token> token = std::nullopt,
+      std::unique_ptr<retry_policy> retry_policy = nullptr);
+
+private:
+    // The root url calculated from base url, prefix and api version. Given a
+    // base url of "/b", an api version "v2" and a prefix of "x/y", the root url
+    // is "/b/v2/x/y/". The root url is prefixed before rest entities used when
+    // making calls to the catalog service
+    ss::sstring root_url() const;
+
+    // Acquires token from catalog API by exchanging credentials
+    ss::future<expected<oauth_token>> acquire_token(retry_chain_node& rtc);
+
+    // Ensures a token is acquired if current token is unset (default state) or
+    // expired. Acquired token is cached for future calls.
+    ss::future<expected<ss::sstring>> ensure_token(retry_chain_node& rtc);
+
+    // Builds the request from supplied builder after validating it, performs
+    // the request with optional payload, and takes care of retrying according
+    // to policy
+    ss::future<expected<iobuf>> perform_request(
+      retry_chain_node& rtc,
+      http::request_builder request_builder,
+      std::optional<iobuf> payload = std::nullopt);
+
+    // TODO sometimes we need to shut down a client, eg in case of transport
+    // errors. If the catalog client is expected to make >1 calls, we need to
+    // acquire a new client after a call fails
+    std::reference_wrapper<http::abstract_client> _client;
```

Comment:
How is this going to work? I feel like it's cleaner if this class owns the client + lifecycle so that it can manage reconstructing a network connection if needed.

Reply:
Ultimately, the design is supposed to be such that the catalog client gets a reference to a client pool, which will be a sharded service so we know it's lifetime, and it acquires clients from the pool as needed. 

Since we do not have a generic pool right now (our current pool implementation is closely related to cloud storage and also lives inside another service), we had to use this setup where http clients are created independently without pooling and then a reference is passed into the catalog client here. 

Owning the client seems to be a good option for now, once we have a pool impl we can switch to using that.

I'll try to switch to a unique ptr instead of a reference for now. We still need an indirection layer so we can get new clients even without a pool if a call fails.

Reply:
I think `catalog_client` is going to have to accept some kind of callable which can create clients on demand, and use that to get clients as needed (most likely in `perform_request`). This should enable acquiring new clients in case of transport errors.

The callable object can be an interface and the pool created later can implement it, for now there can be a factory which creates a new client on each request. I will make the change shortly.

Reply:
I have changed the catalog client to take an http client source, a unique ptr to http client is acquired per call and is shutdown after the call. 

The client source is still taken as a reference, so it should live longer than catalog clients themselves, but the catalog client owns the low level http clients, and can now acquire new ones if needed.